src/shadowcasting.cpp

```diff
@@ -0,0 +1,661 @@
+#include "shadowcasting.h"
+
+#include <list>
+
+#include "enums.h"
+#include "fragment_cloud.h" // IWYU pragma: keep
+#include "line.h"
+
+struct slope {
+    slope( int_least8_t rise, int_least8_t run ) {
+        // Ensure run is always positive for the comparison operators
+        this->run = abs( run );
+        if( run < 0 ) {
+            this->rise = -rise;
+        } else {
+            this->rise = rise;
+        }
+    }
+
+    // We don't need more that 8 bits since the shadowcasting area is not that large,
+    // currently the radius is 60.
+    int_least8_t rise;
+    int_least8_t run;
+};
+
+inline bool operator<( const slope &lhs, const slope &rhs )
+{
+    // a/b < c/d <=> a*d < c*b if b and d have the same sign.
+    return lhs.rise * rhs.run < rhs.rise * lhs.run;
+}
+inline bool operator>( const slope &lhs, const slope &rhs )
+{
+    return rhs < lhs;
+}
+inline bool operator<=( const slope &lhs, const slope &rhs )
+{
+    return !( lhs > rhs );
+}
+inline bool operator>=( const slope &lhs, const slope &rhs )
+{
+    return !( lhs < rhs );
+}
+inline bool operator==( const slope &lhs, const slope &rhs )
+{
+    // a/b == c/d <=> a*d == c*b if b and d have the same sign.
+    return lhs.rise * rhs.run == rhs.rise * lhs.run;
+}
+inline bool operator!=( const slope &lhs, const slope &rhs )
+{
+    return !( lhs == rhs );
+}
+
+template<typename T>
+struct span {
+    span( const slope &s_major, const slope &e_major,
+          const slope &s_minor, const slope &e_minor,
+          const T &value, bool skip_first_row = false ) :
+        start_major( s_major ), end_major( e_major ), start_minor( s_minor ), end_minor( e_minor ),
+        cumulative_value( value ), skip_first_row( skip_first_row ) {}
+    slope start_major;
+    slope end_major;
+    slope start_minor;
+    slope end_minor;
+    T cumulative_value;
+    bool skip_first_row;
+};
+
+/**
+ * Handle splitting the current span in cast_horizontal_zlight_segment and
+ * cast_vertical_zlight_segment to avoid as much code duplication as possible
+ */
+template<typename T, bool( *check )( const T &, const T & ), T( *accumulate )( const T &, const T &, const int & )>
```

Comment:
Would perhaps be clearer if `check` was called something like `is_transparent`?

Reply:
Will fix